packages/mock-account-provider/src/requesters.ts

```diff
@@ -0,0 +1,163 @@
+import fetch from 'node-fetch'
+import * as _ from 'lodash'
+
+export interface GraphqlQueryConfig {
+  resource: string
+  method: string
+  query: string
+  variables: object
+}
+
+export interface GraphqlResponseElement {
+  data: [{ code: string }]
+}
+
+export interface GraphqlResponseArray {
+  [index: number]: GraphqlResponseElement
+}
+
+function testGraphqlElementSuccess(el: GraphqlResponseElement) {
+  _.each(el.data, (v, k) => {
+    if (v.code !== '200') {
+      throw new Error(
+        `graphql response for ${k} contained non-200 code: \n${JSON.stringify(
+          el,
+          null,
+          2
+        )}`
+      )
+    }
+  })
+}
+
+function isGraphqlArray(a: GraphqlResponseElement | GraphqlResponseArray) {
+  if (_.isArray(a)) {
+    return a as GraphqlResponseArray
+  }
+  return a as GraphqlResponseElement
+}
+
+export async function graphqlQuery(options: GraphqlQueryConfig) {
+  return await fetch(options.resource, {
+    method: options.method,
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    body: JSON.stringify({ query: options.query, variables: options.variables })
+  })
+    .then((response) => {
+      if (!response.ok) {
+        throw new Error(
+          `response not ok: ${response.status} ${response.statusText}`
+        )
+      }
+      return response.json()
+    })
+    .then((response: GraphqlResponseElement | GraphqlResponseArray) => {
+      if (isGraphqlArray(response)) {
+        _.each(response, testGraphqlElementSuccess)
+      } else {
+        testGraphqlElementSuccess(response as GraphqlResponseElement)
+      }
+      return response
+    })
+    .catch((e) => {
+      console.log(e)
+      throw e
+    })
+}
+
+export async function createPeer(
+  backendUrl: string,
+  staticIlpAddress: string,
+  outgoingEndpoint: string,
+  assetCode: string,
+  assetScale: number
+): Promise<object> {
+  const createPeerQuery = `
+  mutation CreatePeer ($input: CreatePeerInput!) {
+    createPeer (input: $input) {
+      code
+      success
+      message
+      peer {
+        id
+        asset {
+          code
+          scale
+        }
+        staticIlpAddress
+      }
+    }
+  }
+  `
+  const createPeerInput = {
+    input: {
+      staticIlpAddress,
+      http: {
+        incoming: { authTokens: ['test'] },
+        outgoing: { endpoint: outgoingEndpoint, authToken: 'test' }
+      },
+      asset: {
+        code: assetCode,
+        scale: assetScale
+      }
+    }
+  }
+  return (await graphqlQuery({
+    resource: backendUrl,
+    method: 'post',
+    query: createPeerQuery,
+    variables: createPeerInput
+  })) as unknown as [PeerResponseElement]
+}
+
+export interface PeerResponseElement {
```

Comment:
It is pretty tedious to copy the actual responses I'm getting in here as types, then try to coerce the responses into those types, then use the responses. Placating the type system here is probably 60% of the work. Is there an easier way or is this it?

Reply:
You should be able to generate response types from the schema itself, and then set those types as the return type of this function. You won't need to coerce anything then too. I think code is already being generated by the backend, so you could just use that.

> I was trying to keep the graphql requester generic rather than repeating the fetch calls everywhere. is that a bad idea?

Yeah that seems reasonable. We could use something like the apollo client if this becomes unweildy.